service/gcs/core/gcs/gcs.go

```diff
@@ -406,82 +383,52 @@ func (c *gcsCore) RunExternalProcess(params prot.ProcessParameters, stdioSet *co
 	cmd.SetDir(ociProcess.Cwd)
 	cmd.SetEnv(ociProcess.Env)
 
-	var wg sync.WaitGroup
-	var fileIn, fileOut, fileErr *os.File
-	var master io.ReadWriteCloser
+	var relay *stdio.TtyRelay
 	if params.EmulateConsole {
 		// Allocate a console for the process.
-		// TODO: Should I duplicate the NewConsole functionality outside of
-		// runc to make for better separation between gcscore and runtime?
-		var consolePath string
-		master, consolePath, err = runc.NewConsole()
+		var (
+			consolePath string
+			master      *os.File
+		)
+		master, consolePath, err = stdio.NewConsole()
 		if err != nil {
 			return -1, errors.Wrap(err, "failed to create console for external process")
 		}
+		defer func() {
+			if err != nil {
+				master.Close()
+			}
+		}()
+
 		console, err := c.OS.OpenFile(consolePath, os.O_RDWR, 0777)
 		if err != nil {
 			return -1, errors.Wrap(err, "failed to open console file for external process")
 		}
 		defer console.Close()
 
-		// Begin copying data to and from the console.
-		// In order to ensure master is closed only after it's done being read
-		// from and written to, a sync.WaitGroup is used.
-		if stdioSet.In != nil {
-			wg.Add(1)
-			go func() {
-				io.Copy(master, stdioSet.In)
-				wg.Done()
-			}()
-		}
-		if stdioSet.Out != nil {
-			wg.Add(1)
-			go func() {
-				io.Copy(stdioSet.Out, master)
-				wg.Done()
-			}()
-		}
-
+		relay = stdioSet.NewTtyRelay(master)
 		cmd.SetStdin(console)
 		cmd.SetStdout(console)
 		cmd.SetStderr(console)
 	} else {
-		// Get the underlying files for each stdio socket and use them directly
-		// for the stdio fds for the new process. Close these files as soon as
-		// the process is launched so that only the child process has open
-		// descriptors to these sockets.
-		if stdioSet.In != nil {
-			fileIn, err = stdioSet.In.File()
-			if err != nil {
-				return -1, errors.Wrap(err, "failed to dup stdin socket for command")
-			}
-			defer fileIn.Close()
-			defer stdioSet.In.Close()
-			cmd.SetStdin(fileIn)
-		}
-		if stdioSet.Out != nil {
-			fileOut, err = stdioSet.Out.File()
-			if err != nil {
-				return -1, errors.Wrap(err, "failed to dup stdout socket for command")
-			}
-			defer fileOut.Close()
-			defer stdioSet.Out.Close()
-			cmd.SetStdout(fileOut)
-		}
-		if stdioSet.Err != nil {
-			fileErr, err = stdioSet.Err.File()
-			if err != nil {
-				return -1, errors.Wrap(err, "failed to dup stderr socket for command")
-			}
-			defer fileErr.Close()
-			defer stdioSet.Err.Close()
-			cmd.SetStderr(fileErr)
+		fileSet, err := stdioSet.Files()
+		if err != nil {
+			return -1, errors.Wrap(err, "failed to set cmd stdio")
 		}
+		defer fileSet.Close()
+		defer stdioSet.Close()
```

Comment:
You probably want a ```err == nil``` guard here. The caller of this function closes stdioSet on failure, so if the cmd.Start() fails, you will close this handle twice. Don't think it will be an issue, but the fix is easy enough:
``` go
defer func() { 
    if err == nil {
         stdioSet.Close()
    }
}() 
```